src/librustc_mir/interpret/machine.rs

```diff
@@ -221,17 +231,43 @@ pub trait Machine<'mir, 'tcx>: Sized {
         Ok(())
     }
 
-    /// Called for *every* memory access to determine the real ID of the given allocation.
-    /// This provides a way for the machine to "redirect" certain allocations as it sees fit.
+    /// Called for *every* memory access to determine the real ID of the given
+    /// allocation. This provides a way for the machine to "redirect" certain
+    /// allocations as it sees fit.
     ///
-    /// This is used by Miri to redirect extern statics to real allocations.
+    /// This is used by Miri for two purposes:
+    /// 1.  Redirecting extern statics to real allocations.
+    /// 2.  Creating unique allocation ids for thread locals.
+    ///
+    ///     In Rust, one way for creating a thread local is by marking a static
+    ///     with `#[thread_local]`. On supported platforms this gets translated
+    ///     to a LLVM thread local. The problem with supporting these thread
+    ///     locals in Miri is that in the internals of the compiler they look as
+    ///     normal statics, except that they have the `thread_local` attribute.
+    ///     However, in Miri we want to have a property that each allocation has
+    ///     a unique id and, therefore, for these thread locals we generate a
+    ///     fresh allocation id for each thread.
     ///
     /// This function must be idempotent.
     #[inline]
     fn canonical_alloc_id(_mem: &Memory<'mir, 'tcx, Self>, id: AllocId) -> AllocId {
         id
     }
 
+    /// Called to obtain the `GlobalAlloc` associated with the allocation id.
+    ///
+    /// Miri uses this callback to resolve the information about the original
+    /// thread local static for which `canonical_alloc_id` generated a fresh
+    /// allocation id.
+    #[inline(always)]
+    fn resolve_maybe_global_alloc(
```

Comment:
Does it really not work without this? `canonical_alloc_id` should be enough by returning the correct TLS alloc for the current thread.

Reply:
The core problem for which I cannot find an elegant solution is how to have multiple allocation ids for the same thread local static. The options I though about are:

1.  Duplicating entries of thread-local statics in [`tcx.alloc_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/struct.AllocMap.html). However, `tcx.alloc_map` has an invariant that all statics are deduplicated and I do not think we should start violating it.
2. On the first access of a thread-local static, allocate a fresh thread specific allocation id of a "regular" memory allocation. This could be implemented in at least two ways:

    1. (The current implementation in this PR.) `canonical_alloc_id` only reserves a fresh allocation id, but does not actually allocate memory. The memory gets allocated by `get_raw` and `get_raw_mut` in the same way as for normal statics. The only caveat is that the freshly allocated id does not contain information about the original static and, therefore, we need this additional callback `resolve_maybe_global_alloc` to retrieve information about it.
   2. We could also do the actual allocation in `canonical_alloc_id` because the `memory.alloc_map` is wrapped in `RefCell`. In this way, we would not need the `resolve_maybe_global_alloc` callback. However, as far as I understand, for this to work we would need to make field `alloc_map` and function `get_global_alloc` public.

For me personally, the current implementation felt a little bit more elegant than alternatives. However, I would be happy to change to a different approach.

What is your opinion on this? @bjorn3 @RalfJung @oli-obk

Reply:
I thought `canonicalize_alloc_id` was responsible for returning the `AllocId` corresponding to the allocation of a static. It seems this is done by `Memory::get_global_alloc` instead. Maybe [`eval_const_to_op`](https://github.com/vakaras/rust/blob/3d2d8ea7dcf77d15e669ac9ff0e68ed370f19325/src/librustc_mir/interpret/operand.rs#L539) should instead have a hook to select the alloc id corresponding to the current thread for `#[thread_local]` statics.

Reply:
`canonical_alloc_id` is called first; the result is then fed to `get_global_alloc` if needed (as in, if the resulting `AllocId` is not present in the local `alloc_map`).

https://github.com/rust-lang/rust/blob/1a87c49e33d87955e28fa92a8d59a17264ac6044/src/librustc_mir/interpret/memory.rs#L482-L487

Reply:
> canonical_alloc_id only reserves a fresh allocation id, but does not actually allocate memory. The memory gets allocated by get_raw and get_raw_mut in the same way as for normal statics. The only caveat is that the freshly allocated id does not contain information about the original static and, therefore, we need this additional callback resolve_maybe_global_alloc to retrieve information about it.

Ah, now it makes sense. I think we could merge this PR as it is (+ move this explanation into the source).

Reply:
caveat: we can only merge this PR if miri gets a noop PR that makes it work again without adding new functionality (so do the same change as you did to ConstEvalExtraData and ConstPropExtraData).

Reply:
The `resolve_maybe_global_alloc` calls are positioned in a way such that if you take a reference to a `#[thread_local]` static and send it to another thread, the pointer would resolve to the allocation for the other thread, not the thread that created the reference. `resolve_maybe_global_alloc` needs to be called at the place where the reference is created: `eval_const_to_op`.

Reply:
> The `resolve_maybe_global_alloc` calls are positioned in a way such that if you take a reference to a `#[thread_local]` static and send it to another thread, the pointer would resolve to the allocation for the other thread, not the thread that created the reference. `resolve_maybe_global_alloc` needs to be called at the place where the reference is created: `eval_const_to_op`.

@bjorn3 Do you have an example that should illustrate the issue? I tried the following test and all assertions pass (and as far as I understand, they should):
```rust
#![feature(thread_local)]

use std::thread;

#[thread_local]
static mut A: u8 = 0;
#[thread_local]
static mut B: u8 = 0;
static mut C: u8 = 0;

unsafe fn get_a_ref() -> *mut u8 {
    &mut A
}

fn main() {

    unsafe {
        let x = get_a_ref();
        *x = 5;
        B = 15;
        C = 25;
    }
    
    thread::spawn(|| {
        unsafe {
            assert_eq!(B, 0);
            assert_eq!(C, 25);
            B = 14;
            C = 24;
            let y = get_a_ref();
            assert_eq!(*y, 0);
            *y = 4;
            assert_eq!(*get_a_ref(), 4);
            
        }
    }).join().unwrap();

    unsafe {
        assert_eq!(*get_a_ref(), 5);
        assert_eq!(B, 15);
        assert_eq!(C, 24);
    }
    
}
```

Reply:
I think relevant here is sending pointers to the local thread local across the the thread boundary: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3f26013c3ce170e6733e4bdcc519566d

Reply:
> I think relevant here is sending pointers to the local thread local across the the thread boundary

We should definitely have such a testcase on the Miri side.

Reply:
> I think relevant here is sending pointers to the local thread local across the the thread boundary: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3f26013c3ce170e6733e4bdcc519566d

This test fails with my implementation. Thanks for pointing out!

Reply:
With @oli-obk help, I manage to make the test passing: https://github.com/rust-lang/miri/pull/1284/files#diff-3ee102ac08bbe90033a2dfbf62546499R360-R371

Reply:
@vakaras how does that work? I would have expected this to necessarily affect `eval_const_to_op`, because currently there's just a constant sitting around in the MIR that gets stored in a pointer. Nothing effectful seems to happen, which is the problem tracked in https://github.com/rust-lang/rust/issues/70685.

(The link to the Miri diff unfortunately does not jump to any particular part of that diff.)

When I look at [the MIR here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f43b38bba9b6ca139ae8ae91cd01a305), then I don't see how your approach makes `get_a_ref` return the right pointer *for the current thread*. `canonical_alloc_id` only gets called on accesses, not when the pointer is being created.

Reply:
The Miri diff was supposed to point to the [`access_local`](https://github.com/rust-lang/miri/pull/1284/files#diff-3ee102ac08bbe90033a2dfbf62546499R390) function, which it seems you already found. That function detects that the local being accessed is actually a thread-local static and injects the right allocation id in the same way as `canonical_alloc_id` does.

I agree that this is hacky and we should have a proper solution. However, I am afraid that fixing #70685 will take quite some time and would suggest landing the current version for now and fix #70685 later (unless, of course, you see an alternative design that is less hacky). @RalfJung, does this sound acceptable to you?

Reply:
I hope we can avoid waiting for a fix for #70685. However, even as a temporary stop-gap, I would rather not rely on debug information to execute MIR.

I think a better approach would be what @bjorn3 suggested above: the place where the "weird" global constant enters control of the Miri engine is `eval_const_to_op`, so that's where we have to patch the `AllocId`.

Reply:
I have moved patching to `eval_const_to_op`. Please let me know what you think. @RalfJung @bjorn3 @oli-obk